stetho/src/main/java/com/facebook/stetho/inspector/elements/android/RootsOracle.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+package com.facebook.stetho.inspector.elements.android;
+
+import android.os.Build;
+import android.os.Looper;
+import android.view.View;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static java.lang.Thread.currentThread;
+
+/**
+ * Based on Espresso RootsOracle class:
+ * https://android.googlesource.com/platform/frameworks/testing/+/android-support-test/espresso/
+ * core/src/main/java/android/support/test/espresso/base/RootsOracle.java
+ */
+final class RootsOracle {
+  private static final String WINDOW_MANAGER_IMPL_CLASS = "android.view.WindowManagerImpl";
+  private static final String WINDOW_MANAGER_GLOBAL_CLASS = "android.view.WindowManagerGlobal";
+  private static final String VIEWS_FIELD = "mViews";
+  private static final String GET_DEFAULT_IMPL = "getDefault";
+  private static final String GET_GLOBAL_INSTANCE = "getInstance";
+
+  private boolean initialized;
+  private Object windowManager;
+  private Field viewsField;
+
+  /**
+   * Looks for all Root views
+   */
+  public List<View> findAllRootViews() {
+    if (currentThread() != Looper.getMainLooper().getThread()) {
```

Comment:
Regarding thread affinity, look into implementing the `ThreadBound` interface.

A good way to do this would be to have this class extend from `ThreadBoundProxy`. this now means your constructor needs to take a `ThreadBound` to pass down to the super class. Which means `AndroidDocumentRoot` should take a `ThreadBound` from `AndroidDocumentProvider`, which conveniently has one available from its constructor.